### ISSUE TITLE
No algorithm/protocol available when NCCL_ALGO is set to Tree

### ISSUE THREAD

**Issue by joapolarbear** [NVIDIA/nccl#317]
Hi,

I set NCCL_ALGO=Tree and run the MNist example of horovod on 2 host machines, but it raises the error that no algorithm/protocol is available. I found the problem is related to the following code
```
// src/graph/tuning.cc: ncclSetThresholds()
if (coll != ncclCollAllReduce && a == NCCL_ALGO_TREE) continue;
```
This shows that when the operation is not AllReduce and the algorithm is Tree, the bandwidth is set to 0. In my case, horovod involves a **Broadcast** operation before training. If NCCL_ALGO is set to Tree, since the corresponding bandwidth is 0, the evaluated transmission time is larger than the threshold (1 hour), therefore, no algorithm would be chosen.

I wonder why NCCL does not set the bandwidth for non-AllReduce operations with Tree algorithm. Thanks.

**Comment by kwen2501** [comment]
Hi, thanks for reporting this. You were seeing this error because there is no tree-based implementation for broadcast in NCCL, so `NCCL_ALGO=Tree` does not apply here.

I guess the problem boils down to whether NCCL should use fall-back algorithms when users set it to something that does not exist for certain collectives. This could be _sometimes_ useful when users use mixed collectives and only want to apply `NCCL_ALGO` to one (like AllReduce) than the other. Is this your case?

Btw, NCCL _should_ be able to auto-tune the algorithm and protocol based on the number of nodes & ranks, message size, platform, etc, so hard setting those two environment variables is not always necessary.

**Comment by joapolarbear** [comment]
@kwen2501 Thanks for your reply. I think I will use some fall-back algorithms for those without available algorithm/protocol.

**Comment by sjeaugey** [comment]
@joapolarbear can you explain why you want to force `NCCL_ALGO=TREE` ? This setting is mostly for us to debug performance or tuning issues, but should not be needed by users. If there is a real need for it, please let us know. Also, we could give you a patch to not disable RING algorithms on operations that don't support TREE -- shouldn't be too hard.

**Comment by joapolarbear** [comment]
@sjeaugey  I am working on a program that needs to know some details about the communication operations in NCCL, including the topology and dependency information. We have done the "ring" algorithm part and want to cover the "tree" algorithm. Basically, there is not a real need to use the "tree" algorithm and we just want to design a system that can cover both algorithms.

**Comment by xnning** [comment]
> You were seeing this error because there is no tree-based implementation for broadcast in NCCL, so `NCCL_ALGO=Tree` does not apply here.

Is it still true that there is no tree-based implementation for broadcast? Could you explain why this is the case?

If I am not mistaken, 2-tree algorithm for AllReduce is 2-tree Reduce followed by (maybe pipelined with?) 2-tree Broadcast, can we not implement Reduce as in the first phase for AllReduce, and implement Broadcast as in the second phase for AllReduce?

Thank you.

**Comment by xnning** [comment]
Ah I think I probably get it -- for reduce/broadcast there is no need to have two trees, but a single tree will suffice, i.e., if we got 4 devices,

```
 0
 |
 2
 /\
1 3  
```

a bottom-up will be reduce and a top-down will be broadcast. Is that right?
But unlike two-trees, we probably get less interconnects under working for each step.

**Comment by sjeaugey** [comment]
Yes we could use it for broadcast/reduce but we would have to recompute the trees for each root, and connect ranks, which we don't do at the moment.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
